File: SD/Matlab only/Movie_and_Gif_Maker_from_png_data_color_white_calibration.m
clc
clear
% Script to be run directly in the image folder
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% User parameters
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
target_mp4_file = 'Output.mp4';
target_gif_file = 'Output.gif';
gif_deadtime = 0.1;           
gif_skip = 1;                  
scaling_factor = 0.5;
v_border = 128; 
h_border = 128;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
listing = dir('*.png');
if length(listing) < 3, error('Not enough images.'); end
size_list = length(listing) - rem(length(listing), 3);

% Create folder for full-size frames
if ~exist('RGB_frames', 'dir')
    mkdir('RGB_frames');
end

% --- PASS 1: Calculate Weights ---
disp('Analyzing frames...');
sum_R = 0; sum_G = 0; sum_B = 0; count = 0;
for i = 1:3:size_list
    temp_R = imread(listing(i).name);   temp_R = temp_R(:,:,1);
    temp_G = imread(listing(i+1).name); temp_G = temp_G(:,:,1);
    temp_B = imread(listing(i+2).name); temp_B = temp_B(:,:,1);
    
    sum_R = sum_R + mean(double(temp_R), 'all');
    sum_G = sum_G + mean(double(temp_G), 'all');
    sum_B = sum_B + mean(double(temp_B), 'all');
    count = count + 1;
end
avg_R = sum_R / count; avg_G = sum_G / count; avg_B = sum_B / count;
max_avg = max([avg_R, avg_G, avg_B]);
auto_weights = [max_avg/avg_R, max_avg/avg_G, max_avg/avg_B];

% --- PASS 2: Processing ---
vidfile = VideoWriter(target_mp4_file, 'MPEG-4');
vidfile.FrameRate = 10;
open(vidfile);
gif_counter = 0;

for i = 1:3:size_list
    gif_counter = gif_counter + 1;
    
    % Force 2D read
    R_raw = double(imread(listing(i).name));   R_raw = R_raw(:,:,1);
    G_raw = double(imread(listing(i+1).name)); G_raw = G_raw(:,:,1);
    B_raw = double(imread(listing(i+2).name)); B_raw = B_raw(:,:,1);
    
    % Apply white balance
    R = min(255, R_raw * auto_weights(1));
    G = min(255, G_raw * auto_weights(2));
    B = min(255, B_raw * auto_weights(3));
    
    % Restore borders
    R(1:v_border, :) = R_raw(1:v_border, :); R(end-v_border+1:end, :) = R_raw(end-v_border+1:end, :);
    R(:, 1:h_border) = R_raw(:, 1:h_border); R(:, end-h_border+1:end) = R_raw(:, end-h_border+1:end);
    G(1:v_border, :) = G_raw(1:v_border, :); G(end-v_border+1:end, :) = G_raw(end-v_border+1:end, :);
    G(:, 1:h_border) = G_raw(:, 1:h_border); G(:, end-h_border+1:end) = G_raw(:, end-h_border+1:end);
    B(1:v_border, :) = B_raw(1:v_border, :); B(end-v_border+1:end, :) = B_raw(end-v_border+1:end, :);
    B(:, 1:h_border) = B_raw(:, 1:h_border); B(:, end-h_border+1:end) = B_raw(:, end-h_border+1:end);
    
    % Create full-size frame
    full_frame = uint8(cat(3, R, G, B));
    
    % Save to RGB_frames folder
    imwrite(full_frame, fullfile('RGB_frames', sprintf('frame_%04d.png', gif_counter)));
    
    % Resize for video/GIF
    target_size = [floor(size(R_raw, 1) * scaling_factor), floor(size(R_raw, 2) * scaling_factor)];
    R_s = uint8(imresize(R, target_size, 'nearest'));
    G_s = uint8(imresize(G, target_size, 'nearest'));
    B_s = uint8(imresize(B, target_size, 'nearest'));
    frame_scaled = cat(3, R_s, G_s, B_s);
    
    % Write Video
    writeVideo(vidfile, frame_scaled);
    
    % GIF generation
    [imind, map] = rgb2ind(frame_scaled, 256);
    if i == 1
        imwrite(imind, map, target_gif_file, 'gif', 'Loopcount', inf, 'DelayTime', gif_deadtime);
    elseif rem(gif_counter, gif_skip) == 0
        imwrite(imind, map, target_gif_file, 'gif', 'WriteMode', 'append', 'DelayTime', gif_deadtime);
    end
end
close(vidfile);
disp('Conversion complete.');

File: SD/Python/Movie_and_Gif_Maker_from_png_data_color_white_calibration.py
import cv2
import numpy as np
import os
import glob
import imageio.v3 as iio

# User parameters
target_mp4_file = 'Output.mp4'
target_gif_file = 'Output.gif'
gif_deadtime = 100  # imageio uses milliseconds (0.1s = 100ms)
scaling_factor = 0.5
v_border = 128
h_border = 128

# Create folder for full-size frames
output_folder = 'RGB_frames'
if not os.path.exists(output_folder):
    os.makedirs(output_folder)

# Get list of images - verify these are in the same folder as the script
files = sorted(glob.glob('*.png'))
if len(files) < 3:
    raise ValueError(f'Not enough images found. Found {len(files)}, need at least 3.')

size_list = len(files) - (len(files) % 3)

# --- PASS 1: Calculate Weights ---
print('Analyzing frames...')
sum_R, sum_G, sum_B = 0.0, 0.0, 0.0
count = 0
for i in range(0, size_list, 3):
    sum_R += np.mean(cv2.imread(files[i], cv2.IMREAD_GRAYSCALE))
    sum_G += np.mean(cv2.imread(files[i+1], cv2.IMREAD_GRAYSCALE))
    sum_B += np.mean(cv2.imread(files[i+2], cv2.IMREAD_GRAYSCALE))
    count += 1

avg_R, avg_G, avg_B = sum_R/count, sum_G/count, sum_B/count
max_avg = max([avg_R, avg_G, avg_B])
weights = [max_avg/avg_R, max_avg/avg_G, max_avg/avg_B]

# --- PASS 2: Processing ---
h, w = cv2.imread(files[0], cv2.IMREAD_GRAYSCALE).shape
target_size = (int(w * scaling_factor), int(h * scaling_factor))
vid = cv2.VideoWriter(target_mp4_file, cv2.VideoWriter_fourcc(*'mp4v'), 10.0, target_size)

gif_frames = []

for i in range(0, size_list, 3):
    # Read raw
    R_raw = cv2.imread(files[i], cv2.IMREAD_GRAYSCALE).astype(np.float32)
    G_raw = cv2.imread(files[i+1], cv2.IMREAD_GRAYSCALE).astype(np.float32)
    B_raw = cv2.imread(files[i+2], cv2.IMREAD_GRAYSCALE).astype(np.float32)
    
    # White Balance
    R = np.clip(R_raw * weights[0], 0, 255)
    G = np.clip(G_raw * weights[1], 0, 255)
    B = np.clip(B_raw * weights[2], 0, 255)
    
    # Restore borders
    for channel, raw in [(R, R_raw), (G, G_raw), (B, B_raw)]:
        channel[:v_border, :] = raw[:v_border, :]
        channel[-v_border:, :] = raw[-v_border:, :]
        channel[:, :h_border] = raw[:, :h_border]
        channel[:, -h_border:] = raw[:, -h_border:]
    
    # Save full size frame
    full_frame = cv2.merge([B, G, R]).astype(np.uint8)
    cv2.imwrite(os.path.join(output_folder, f'frame_{i//3:04d}.png'), full_frame)
    
    # Prepare for Video/GIF (Resized)
    frame_s = cv2.resize(full_frame, target_size, interpolation=cv2.INTER_NEAREST)
    vid.write(frame_s)
    gif_frames.append(cv2.cvtColor(frame_s, cv2.COLOR_BGR2RGB))

# Finalize outputs
vid.release()
iio.imwrite(target_gif_file, gif_frames, duration=gif_deadtime, loop=0)
print('Conversion complete. MP4 and GIF saved.')